api_helper.py

```diff
@@ -1,17 +1,99 @@
+    def __init__(self, connection, cursor):
+        self.connection = connection
+        self.cursor = cursor
+
+    @log_time
+    def fetch_data(self, url, country=None):
         try:
             if country is not None:
-                url=url+f"country={country}"
+                url = url + f"?country={country}"
             response = requests.get(url)
             if response.status_code == 200:
-                return response.json()        
+                return response.json()
+        except requests.RequestException as e:
+            logging.error(f"Failed to fetch data for {url}: {e}")
+            return None
+
+    def close_connection(self):
+        if self.cursor is not None and not self.cursor.closed:
+            self.cursor.close()
+        if self.connection is not None and  self.connection.closed == 0:
+            self.connection.close()
+
+    @log_time
+    def insert_data(self, data_to_insert):
+        try:
+            for table_name, data in data_to_insert.items():
+                if table_name == "country":
+                    for country in data:
+                        self._insert_generic("country", {"country_name": country})
+                elif table_name == "university":
+                    for country, universities in data.items():
+                        country_id = self._insert_generic("country", {"country_name": country})
+                        for university in universities:
+                            university_data = {
+                                "university_name": university['name'],
+                                "country_id": country_id
+                            }
+                            self._insert_generic("university", university_data)
+                elif table_name == "person":
+                    for person_data in data:
+                        flattened_data = self._flatten_person_data(person_data)
+                        country = flattened_data['country']
+                        country_id = self._insert_generic("country", {"country_name": country})
+                        flattened_data['country_id'] = country_id
+                        del flattened_data['country']
+                        self._insert_generic("person", flattened_data)
+            yield "Data insertion completed."
+        except Error as e:
+            self.connection.rollback()  # Rollback the transaction
+            logging.error(f"Error while inserting data: {e}")
+            yield f"Error occurred: {e}"
+        except psycopg2.InterfaceError as e:
+            logging.error(f"InterfaceError occurred: {e}")
+
+    def _insert_generic(self, table_name, data):
+        try:
+            columns = data.keys()
+            values = tuple(data.values())
+            placeholders = ', '.join(['%s'] * len(values))
+            query = f"INSERT INTO {table_name} ({', '.join(columns)}) VALUES ({placeholders}) ON CONFLICT DO NOTHING RETURNING {table_name}_id"
```

Comment:
instead of `ON CONFLICT DO NOTHING RETURNING` use `ON CONFLICT DO update`
because we want to maintain latest arrived data

Reply:
I rewrite the code conflict code as below
![image](https://github.com/DGMahendra/api_project/assets/161186439/a6b9ec37-5716-4a46-b11d-f0efb829d5db)